From e1b7c57a8b9cdf55ab56cb2b8f37996a4315a967 Mon Sep 17 00:00:00 2001
From: Chris Kay <chris.kay@arm.com>
Date: Tue, 16 Jan 2024 11:53:35 +0000
Subject: [PATCH 718/954] build: add `--no-warn-rwx-segments` when linking with
 GCC

This flag is currently not included when linking with BFD via GCC.
Without it, builds can fail when linking via GCC.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git e91b5394e34a9dbd1a426aee4288396f0cefdd4e build: add `--no-warn-rwx-segments` when linking with GCC]
Change-Id: Id37e05f6fb4eea4620861eb1bc8668c41f21ba20
Signed-off-by: Chris Kay <chris.kay@arm.com>
(cherry picked from commit 86e489c1903c72b76c6ff3b3ff6b75774c132e59)
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/422806
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Yann GAUTIER <yann.gautier@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Maxime MERE <maxime.mere@foss.st.com>
Tested-by: Maxime MERE <maxime.mere@foss.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/438768
Domain-Review: Maxime MERE <maxime.mere@foss.st.com>
---
 Makefile                     | 1 +
 make_helpers/build_macros.mk | 4 +---
 2 files changed, 2 insertions(+), 3 deletions(-)

diff --git a/Makefile b/Makefile
index fc9f47125a..b3c1f29e48 100644
--- a/Makefile
+++ b/Makefile
@@ -350,6 +350,7 @@ ifneq ($(findstring armlink,$(notdir $(LD))),)
 # LD = gcc (used when GCC LTO is enabled)
 else ifneq ($(findstring gcc,$(notdir $(LD))),)
 	# Pass ld options with Wl or Xlinker switches
+	TF_LDFLAGS		+=	$(call ld_option,-Xlinker --no-warn-rwx-segments)
 	TF_LDFLAGS		+=	-Wl,--fatal-warnings -O1
 	TF_LDFLAGS		+=	-Wl,--gc-sections
 
diff --git a/make_helpers/build_macros.mk b/make_helpers/build_macros.mk
index 36ecb32066..e3194a8891 100644
--- a/make_helpers/build_macros.mk
+++ b/make_helpers/build_macros.mk
@@ -113,9 +113,7 @@ endef
 
 # Convenience function to check for a given linker option. An call to
 # $(call ld_option, --no-XYZ) will return --no-XYZ if supported by the linker
-define ld_option
-	$(shell if $(LD) $(1) -v >/dev/null 2>&1; then echo $(1); fi )
-endef
+ld_option = $(shell $(LD) $(1) -Wl,--version >/dev/null 2>&1 || $(LD) $(1) -v >/dev/null 2>&1 && echo $(1))
 
 # Convenience function to check for a given compiler option. A call to
 # $(call cc_option, --no-XYZ) will return --no-XYZ if supported by the compiler
-- 
2.53.0

